ccp/src/hashrate/handler.rs

```diff
@@ -87,4 +82,16 @@ impl HashrateHandler {
     pub(crate) fn sliding_hashrate(&self) -> &SlidingHashrate {
         self.sliding_collector.hashrate()
     }
+
+    pub(crate) fn handle_epoch(&self, epoch: EpochParameters) -> HResult<()> {
+        let mut guard = self.collector.lock().unwrap();
+        if let EpochObservation::EpochChanged {
+            prev_epoch_hashrate,
+        } = guard.observe_epoch(epoch)
+        {
+            self.saver.save_hashrate_previous(prev_epoch_hashrate)?;
```

Comment:
The guard is still hold here.  Is it really needed?

Reply:
You are saving the current hashrate + cleaning the things up. Doesn't this worth a lock?

Reply:
Then it should be lock on `saver`.

Reply:
After some discussion in slack I am stating the lock granularity here should be addressed by VM-589.